### PR TITLE
KAFKA-17527: Fix NPE for null RecordContext

### PR DESCRIPTION
**Pull request by mjsax**
\cc @cmccabe -- this is a blocker for 3.9 and must be cherry-picked

### CODE REVIEW

**Comment by mjsax** [comment]
\cc @sebastienviale @loicgreffier

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java]: Given what the callback is executed on the producer sender thread, I think we need to make a deep copy here. Otherwise, the information inside the `RecordContext` object might change under the hood.

File: streams/src/main/java/org/apache/kafka/streams/errors/ErrorHandlerContext.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/errors/ErrorHandlerContext.java:49]: Could you be more explicit here?
The error handler context is used in processing exception handler, production exception handler, and deserialization exception handler. It is not immediately clear that this sentence only applies to the production exception handler.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java]: Isn't the issue here that `context` might be `null`?
See https://github.com/apache/kafka/blob/a4ea9aec73a3da470d079bad82d92810cac49d55/streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java#L143

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java]: See comment below.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java]: See comment below.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java]: I am not sure, whether this makes a difference. Once you get the reference with `context.recordContext()` the record context will not change anymore since it is an immutable object.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java]: To avoid code duplication, I think it is better to do this when you verify that `context` is `null` in `send()`.

**Review by cadonna — COMMENTED** [review]
Thanks for the PR @mjsax !

Here my feedback!

File: streams/src/main/java/org/apache/kafka/streams/errors/ErrorHandlerContext.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/errors/ErrorHandlerContext.java:49]: > when writing into a changelog topic

Seems pretty clear to me that it's not on the read nor processing path.  -- The statements above area also not more precise. Happy to improve it, but not sure how? Do you have a more concrete suggestions what you want to see (maybe also for the existing statements)?

While I am always in favor of good JavaDocs, also wondering if stuff might out date w/o getting updated when we change things in the runtime. Hence, having some "wiggle room" might actually benefit us?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java]: Not sure how this would work? -- Yes, we need to check if  `context == null`, and for this case set `recordContext = null` instead of `recordContext = context.recordContext()`. But not sure how I can avoid the `recordContext` null check?

Let me extract into a helper method to avoid code duplication though.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:257]: Wrapping the whole `Callback` with a try-catch to avoid swallowing exceptions.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java:295]:
```suggestion
                        context, // ok as-is; `null` check done inside `recordSendError(...)`
```

File: streams/src/main/java/org/apache/kafka/streams/errors/ErrorHandlerContext.java
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/errors/ErrorHandlerContext.java:49]: The issue probably starts already in the class javadoc since it only mentions failed processing. 
I would extend that javadoc and describe all three cases in which the error context is used and give names to the cases. Something like deserialization exception handling, processing exception handling, and production exception handling. Then here we could write during production exception handling when writing into a changelog topic no input topic name is available.

We can do that in a follow-up PR. Let us not block this PR on this.

Comments are always hard to keep up-to-date. That is the reason, I do not like inline comments because they eventually start to lie at some point. Also Javadocs are challenging to keep up-to-date, but it is worth the pain, IMO.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java]: Could you remove the inline comment, please 🙏 ?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java]: Could you please remove the inline comment?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java]: Please remove.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java]: Then maybe, it makes sense to have an overload for the constructor instead of an inline comment.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java:1607]: A variable `notAvailableContext = null` would be better than the inline comment.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java:1671]: See my comment above.

**Review by cadonna — APPROVED** [review]
Thanks for the updates!

I had some comments that you do not need to fix in this PR.

I am currently soaking the fix.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java:1607]: I disagree -- would just raise the question, "why not available" when reading the code, and does not really explain that pass in `null` here is the actual test condition.

File: streams/src/main/java/org/apache/kafka/streams/errors/ErrorHandlerContext.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/errors/ErrorHandlerContext.java:49]: In general I agree, but I would not be too strict about in-line comments -- they can be very helpful to explain important context, which cannot be expressed by code itself IMHO.

I did push another smaller change that I think is valuable for 3.9 release, but will leave it to follow up PRs to improve this further fro 4.0+ release.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java]: Matter of personal taste -- I personally don't like constructor overloads too much. -- It also does not explain when you need to pass the context and when not... only a comment can explain this.

However, we can still pass in the `context` here, too -- it won't be used, but it won't hurt either...

File: streams/src/main/java/org/apache/kafka/streams/errors/ErrorHandlerContext.java
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/errors/ErrorHandlerContext.java:49]: > they can be very helpful to explain important context, which cannot be expressed by code itself IMHO.

I disagree on this. You can express a lot with code and meaningful names which are better maintanable than inline comments. Of course, there are exceptions.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java:1607]: Do you think "pass `null` context for testing" is clearer?

The test name is `shouldNotFailIfContextIsNotAvailableOnSerializationError`, I think it is quite clear why the context is not available and that this is the test condition.

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java:1607]: Yes, I think it's clearer :) (otherwise, I would have done it differently). It highlights where the `null` the test name indicates already is exactly passed.

Of course, using `nullContext = null` and pass `nullContext` would also do the trick. It's just that both are equally good for this case IMHO. -- There are other cases for which I would prefer a variable, eg, `UNKNOWN = -1`, and pass `UNKNOWN` instead of `-1` with a comment. However, for the unknown case, I don't care how unknowns is encoded, ie, it could also be `-2` and thus using a variable is better. For `null` vs `nullContext` there is just no read difference, because `null` can only be `null`.

But I guess we are bikeshedding now.

File: streams/src/main/java/org/apache/kafka/streams/errors/ErrorHandlerContext.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/errors/ErrorHandlerContext.java:49]: > You can express a lot with code and meaningful names which are better maintanable than inline comments. Of course, there are exceptions.

We are on the same page. I did say, "In general I agree". And I believe, this case in one of the exception you called out :)

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `3.9` branch.

**Pull request merged**
